### DISCUSSION TITLE
tutorials/voyage-part2/08-stub-subgraph

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#144] [course feedback]
# tutorials/voyage-part2/08-stub-subgraph

https://www.apollographql.com/tutorials/voyage-part2/08-stub-subgraph

**Comment by sheilaklotz** [comment]
When, in chapter 3, I converted my monolith to a subgraph, it convert it to version Federation 2.8. But the stub subgraph does not work if I user federation 2.8. It does not recognize the directives @shareable and @inaccssible. So, I used version 2.7.

But now, when I try to restart the router, I get messages about incompatibility, since each subgraph has a different version. How to resolve this situation. Maybe it is possible to set the federation to 2.7 version when I do the convert. 

Please, advise. Thank you

**Reply by mabuyo** [reply]
Hey there @sheilaklotz ! Can you paste in the messages about incompatibility you're getting? Are they blocking the router from starting up?

**Reply by mabuyo** [reply]
I believe the course is also using 2.7 at the moment and we haven't updated to 2.8 yet. You can try changing the build pipeline federation version for the graph in Studio by:

1. Go to your variant's Settings page, then find the General > Build Pipeline section.
2. Click Upgrade.
3. In the dialog that appears, select a Federation version 2.7 from the dropdown and click Save.

**Reply by sheilaklotz** [reply]
ok, do not changing federation to 2.8 in any file, I was able to set rover dev, without the compatibility error
supergraph config loaded successfully

==> Watching ./router/router-config.yaml for changes
warning: Do not run this command in production! It is intended for local development only.
starting a session with the 'accounts' subgraph
==> Watching ./subgraph-accounts/schema.graphql for changes
starting a session with the 'monolith' subgraph
composing supergraph with Federation v2.7.0
WARN: telemetry.instrumentation.spans.mode is currently set to 'deprecated', either explicitly or via defaulting. Set telemetry.instrumentation.spans.mode explicitly in your router.yaml to 'spec_compliant' for log and span attributes that follow OpenTelemetry semantic conventions. This option will be defaulted 
to 'spec_compliant' in a future release and eventually removed altogether
==> your supergraph is running! head to http://localhost:4000 to query your supergraph
successfully composed after adding the 'monolith' subgraph
==> Watching ./monolith/schema.graphql for changes

I suppose I am good to go

**Comment by sheilaklotz** [comment]
I am using router version 1.49 instead 1.46 like in the course, since my supergraph is federation 2.8


​--------------------------------------------------------------------------------------
Sheila Klotz
Kyndryl Consult - Cloud Export
SHARED - Microsoft SQL Server Database Administr​ator SME
Phone +55 19 3201 4319
***@***.***
GMT -3
​--------------------------------------------------------------------------------------
________________________________
De: Sheila Klotz ***@***.***>
Enviado: quarta-feira, 16 de outubro de 2024 11:32
Para: apollographql-education/content-feedback ***@***.***>; apollographql-education/content-feedback ***@***.***>
Cc: Mention ***@***.***>
Assunto: RE: [EXTERNAL] Re: [apollographql-education/content-feedback] tutorials/voyage-part2/08-stub-subgraph (Discussion #144)

In the moment I am trying do not change nothing in the files for 2.8, letting all 2.7 to see if the error persists. So, my supergraph uses federation 2.8, but my subgraphos are using federation 2.7.

Does it work?


​--------------------------------------------------------------------------------------
Sheila Klotz
Kyndryl Consult - Cloud Export
SHARED - Microsoft SQL Server Database Administr​ator SME
Phone +55 19 3201 4319
***@***.***
GMT -3
​--------------------------------------------------------------------------------------
________________________________
De: Michelle Mabuyo ***@***.***>
Enviado: quarta-feira, 16 de outubro de 2024 11:27
Para: apollographql-education/content-feedback ***@***.***>
Cc: Sheila Klotz ***@***.***>; Mention ***@***.***>
Assunto: [EXTERNAL] Re: [apollographql-education/content-feedback] tutorials/voyage-part2/08-stub-subgraph (Discussion #144)

I believe the course is also using 2. 7 at the moment and we haven't updated to 2. 8 yet. You can try changing the build pipeline federation version for the graph in Studio by: Go to your variant's Settings page, then find the General > Build


I believe the course is also using 2.7 at the moment and we haven't updated to 2.8 yet. You can try changing the build pipeline federation version for the graph in Studio by:

  1.  Go to your variant's Settings page, then find the General > Build Pipeline section.
  2.  Click Upgrade.
  3.  In the dialog that appears, select a Federation version 2.7 from the dropdown and click Save.

**Reply by mabuyo** [reply]
Got it, thanks for the extra context. It looks like router v1.49 supports at most federation 2.8.1. If your build pipeline is set to 2.8 in Studio, that should be okay. And subgraphs should also be okay to stay at 2.7 (according [to this docs article](https://www.apollographql.com/docs/graphos/platform/graph-management/updates#3-update-subgraph-schemas-as-needed)). If you could let me know the error you're seeing, I can try to troubleshoot further!

**Comment by monish001** [comment]
`rover dev` is not working for me. Getting below error. Any clue?

```
** Notice: Changes in This Release! **
This version includes significant updates to the `rover dev` command.
We highly recommend reviewing the updated documentation to ensure a smooth experience.
Read more: https://www.apollographql.com/docs/rover/commands/dev

warning: Do not run this command in production! It is intended for local development only.

merging supergraph schema files
supergraph config loaded successfully
error: Error occurred when composing supergraph
Error resolving subgraphs:
Unable to resolve subgraphs.
accounts: Could not find schema file (./subgraph-accounts/schema.graphql) relative to (/Users/username/Documents/GitHub/odyssey-voyage-II-server/router) for subgraph `accounts`
monolith: Could not find schema file (./monolith/schema.graphql) relative to (/Users/username/Documents/GitHub/odyssey-voyage-II-server/router) for subgraph `monolith`
```

**Reply by lizhennessy** [reply]
Hey there @monish001 ! Which directory were you running the `rover dev` command in? From the error it looks like the relative paths to each subgraph schema might not have been quite what the command expected! Can you try from the root of the `odyssey-voyage-II-server` directory?

**Reply by monish001** [reply]
Thank you @lizhennessy for checking out my comment.

I am running `rover dev` from `odyssey-voyage-II-server` directory. Please see below. I have included output of the `ls`.
```
username@hostname odyssey-voyage-II-server % ls
CODEOWNERS              final                   monolith                services                subgraph-template
README.md               log.log                 router                  subgraph-accounts
username@hostname odyssey-voyage-II-server % rover dev --supergraph-config ./router/supergraph-config.yaml \
   --router-config ./router/router-config.yaml

** Notice: Changes in This Release! **
This version includes significant updates to the `rover dev` command.
We highly recommend reviewing the updated documentation to ensure a smooth experience.
Read more: https://www.apollographql.com/docs/rover/commands/dev

warning: Do not run this command in production! It is intended for local development only.

merging supergraph schema files
supergraph config loaded successfully
error: Error occurred when composing supergraph
Error resolving subgraphs:
Unable to resolve subgraphs.
accounts: Could not find schema file (./subgraph-accounts/schema.graphql) relative to (/Users/username/Documents/GitHub/odyssey-voyage-II-server/router) for subgraph `accounts`
monolith: Could not find schema file (./monolith/schema.graphql) relative to (/Users/username/Documents/GitHub/odyssey-voyage-II-server/router) for subgraph `monolith`
```

**Reply by lizhennessy** [reply]
Thanks for the added detail @monish001 ! What happens if you go into the `supergraph-config.yaml` file and modify the path to each subgraph? So for instance, changing each path to the following to go up one directory:

```
subgraphs:
  monolith:
    routing_url: http://localhost:4001
    schema:
      file: ../monolith/schema.graphql
  accounts:
    routing_url: http://localhost:4002
    schema:
      file: ../subgraph-accounts/schema.graphql
```

Since we're running the command in the root of the project file, what we've started with should work, but I'm curious if there's a directory issue here. Otherwise, I'm going to run through the code to make sure that something else isn't happening to prevent your command from working!

**Reply by lizhennessy** [reply]
I just ran through the code and was able to run `rover dev` in the root of the `odyssey-voyage-ii-server` directory - so the process was able to find both my subgraphs, `monolith` and `subgraph-accounts` with the original file path to the schema. The idea is that the path to each schema file is relative to the directory where we run `rover dev`; so in this case, from the root, the path would be correct. 🤔 What operating system are you running this command on? It might also be useful to check the Rover version you're running with `rover --version` anywhere in the terminal - I'm on `0.25.0` for instance.

**Reply by monish001** [reply]
Thank you @lizhennessy for the set of inputs.

Above suggestions seem to have helped! I will update back in case i see any more issues.

(1)
I tried change you suggested in the file `supergraph-config.yaml`, it is started working now.
```
username@hostname odyssey-voyage-II-server % rover dev --supergraph-config ./router/supergraph-config.yaml \
   --router-config ./router/router-config.yaml

** Notice: Changes in This Release! **
This version includes significant updates to the `rover dev` command.
We highly recommend reviewing the updated documentation to ensure a smooth experience.
Read more: https://www.apollographql.com/docs/rover/commands/dev

warning: Do not run this command in production! It is intended for local development only.

merging supergraph schema files
supergraph config loaded successfully
starting a session with the 'accounts' subgraph
==> Watching /Users/username/Documents/GitHub/odyssey-voyage-II-server/subgraph-accounts/schema.graphql for changes
starting a session with the 'monolith' subgraph
==> Watching /Users/username/Documents/GitHub/odyssey-voyage-II-server/monolith/schema.graphql for changes
composing supergraph with Federation 2.7.0
==> Watching ./router/router-config.yaml for changes
==> Attempting to start router at http://localhost:4000.
WARN: telemetry.instrumentation.spans.mode is currently set to 'deprecated', either explicitly or via defaulting. Set telemetry.instrumentation.spans.mode explicitly in your router.yaml to 'spec_compliant' for log and span attributes that follow OpenTelemetry semantic conventions. This option will be defaulted to 'spec_compliant' in a future release and eventually removed altogether
==> Health check exposed at http://127.0.0.1:8088/health
==> Your supergraph is running! head to http://localhost:4000 to query your supergraph
```

(2)
I am MacOS.

(3)
Rover version:
```
username@hostname odyssey-voyage-II-server % rover --version
Rover 0.27.2
```

**Reply by lizhennessy** [reply]
Great! I'm so glad that was able to resolve the issue. Of course do let us know if anything else out of the ordinary pops up. :)

**Comment by tohure** [comment]
When I run the command (rover dev . . . .), it asks me questions like where I'm running the subgraph and its name. Should I enter them one by one?:

- what URL is your subgraph running on?: . . . . .
- what is the name of this subgraph? [router]: . . . .

**Reply by lizhennessy** [reply]
Hey there @tohure - you're seeing these questions even after including the `--supergraph-config` and `--router-config` arguments? 🤔 (Are your subgraphs also running at this time?)

**Reply by tohure** [reply]
Yes, I just checked and my subgraphs are running (monolith and accounts) and both parameters mentioned.
![Screenshot 2025-05-30 at 12 05 03 PM](https://github.com/user-attachments/assets/f44cac79-8325-4548-9c36-80d1829f70e2)

**Reply by tohure** [reply]
Thanks @lizhennessy, reviewing the previous answers, I changed the command slightly by running it in the same "Rover folder". I also changed the configuration yaml.

`rover dev --supergraph-config supergraph-config.yaml --router-config router-config.yaml`

```
federation_version: =2.10.0
    subgraphs:
      monolith:
        routing_url: http://localhost:4001
        schema:
          file: ../monolith/schema.graphql
      accounts:
        routing_url: http://localhost:4002
        schema:
          file: ../subgraph-accounts/schema.graphql
```

![Screenshot 2025-05-30 at 1 22 44 PM](https://github.com/user-attachments/assets/8ee32afd-50f1-4c49-bcaf-5c28973c1107)

Regards!
